pkg/controller/datavolume/controller-base.go

```diff
@@ -323,18 +341,115 @@ func (r ReconcilerBase) syncUpdate(log logr.Logger, syncRes *dataVolumeSyncResul
 	return nil
 }
 
-func (r ReconcilerBase) handlePrePopulation(dv *cdiv1.DataVolume, pvc *corev1.PersistentVolumeClaim) {
+func (r *ReconcilerBase) handleStaticVolume(syncRes *dataVolumeSyncResult, log logr.Logger) error {
+	if _, ok := syncRes.dvMutated.Annotations[cc.AnnCheckStaticVolume]; !ok {
+		return nil
+	}
+
+	if syncRes.pvc == nil {
+		volumes, err := r.getAvailableVolumesForDV(syncRes, log)
+		if err != nil {
+			return err
+		}
+
+		if len(volumes) == 0 {
+			log.Info("No PVs for DV")
+			return nil
+		}
+
+		if err := r.handlePvcCreation(log, syncRes, func(_ *cdiv1.DataVolume, pvc *corev1.PersistentVolumeClaim) error {
+			bs, err := json.Marshal(volumes)
+			if err != nil {
+				return err
+			}
+			cc.AddAnnotation(pvc, cc.AnnPersistentVolumeList, string(bs))
+			return nil
+		}); err != nil {
+			return err
+		}
+
+		// set result to make sure callers don't do anything else in sync
+		syncRes.result = &reconcile.Result{}
+		return nil
+	}
+
+	volumeAnno, ok := syncRes.pvc.Annotations[cc.AnnPersistentVolumeList]
+	if !ok {
+		// etiher did not create the PVC here OR bind to expected PV succeeded
+		return nil
+	}
+
+	if syncRes.pvc.Spec.VolumeName == "" {
+		// set result to make sure callers don't do anything else in sync
+		syncRes.result = &reconcile.Result{}
+		return nil
+	}
+
+	var volumes []string
+	if err := json.Unmarshal([]byte(volumeAnno), &volumes); err != nil {
+		return err
+	}
+
+	for _, v := range volumes {
+		if v == syncRes.pvc.Spec.VolumeName {
+			pvcCpy := syncRes.pvc.DeepCopy()
+			// handle as "populatedFor" going foreward
+			cc.AddAnnotation(pvcCpy, cc.AnnPopulatedFor, syncRes.dvMutated.Name)
+			delete(pvcCpy.Annotations, cc.AnnPersistentVolumeList)
+			if err := r.updatePVC(pvcCpy); err != nil {
+				return err
+			}
+			syncRes.pvc = pvcCpy
+			return nil
+		}
+	}
+
+	// delete the pvc and hope for better luck...
+	pvcCpy := syncRes.pvc.DeepCopy()
+	if err := r.client.Delete(context.TODO(), pvcCpy, &client.DeleteOptions{}); err != nil {
+		return err
+	}
+
+	syncRes.pvc = pvcCpy
+
+	return fmt.Errorf("DataVolume bound to unexpected PV %s", syncRes.pvc.Spec.VolumeName)
+}
+
+func (r *ReconcilerBase) getAvailableVolumesForDV(syncRes *dataVolumeSyncResult, log logr.Logger) ([]string, error) {
+	pvList := &corev1.PersistentVolumeList{}
+	fields := client.MatchingFields{claimRefField: claimRefIndexKeyFunc(syncRes.dv.Namespace, syncRes.dv.Name)}
+	if err := r.client.List(context.TODO(), pvList, fields); err != nil {
+		return nil, err
+	}
+	var pvNames []string
+	for _, pv := range pvList.Items {
+		// TODO - should we be more specific here, like do the actual matching in pv controller?
```

Comment:
Maybe something from https://pkg.go.dev/k8s.io/component-helpers@v0.26.1/storage/volume
can help here?

Reply:
unfortunately that package is not as fleshed out in k8s 23 but using it anyway